projects/driver_controls/src/throttle.c

```diff
@@ -0,0 +1,144 @@
+// prv_flag_update_callback is called to set reading_updated_flag accordingly which determines if
+// the readings are up to date. A periodic safety check, prv_raise_event_timer_callback, checks
+// for this flag and if ok, it updates the position of the pedal in storage and
+// raises the events related to the pedal's position (braking, coasting, and accelerating zones).
+// The reading_ok_flag basically holds the actual state of reading_updated_flag when it is reset.
+// It also verifies if the second channel is in sync with the main channel to ensure the validity
+// of the data.
+// If the data turns out to be stale or channels aren't synched, a timeout event will be raised.
+// The throttle_get_position function simply reads the position from storage.
+#include "throttle.h"
+#include <string.h>
+#include "event_queue.h"
+#include "input_event.h"
+
+// Scales a reading to a measure out of 12 bits based on the given range (min - max).
+static uint16_t prv_scale_reading(int16_t reading, int16_t max, int16_t min) {
+  return (1 << 12) * (reading - min) / (max - min);
+}
+
+static uint16_t prv_get_numerator(int16_t reading, ThrottleChannel channel,
+                                  ThrottleStorage *storage) {
+  return prv_scale_reading(
+      reading, storage->calibration_data->line_of_best_fit[channel][THROTTLE_THRESH_MAX],
+      storage->calibration_data->line_of_best_fit[channel][THROTTLE_THRESH_MIN]);
+}
+
+static uint16_t prv_get_numerator_zone(int16_t reading_main, ThrottleZone zone,
+                                       ThrottleStorage *storage) {
+  return prv_scale_reading(
+      reading_main, storage->calibration_data->zone_thresholds_main[zone][THROTTLE_THRESH_MAX],
+      storage->calibration_data->zone_thresholds_main[zone][THROTTLE_THRESH_MIN]);
+}
+
+static bool prv_reading_within_zone(int16_t reading_main, ThrottleZone zone,
+                                    ThrottleStorage *storage) {
+  return (
+      (reading_main < storage->calibration_data->zone_thresholds_main[zone][THROTTLE_THRESH_MAX]) &&
+      (reading_main >= storage->calibration_data->zone_thresholds_main[zone][THROTTLE_THRESH_MIN]));
+}
+
+// Returns true if the channels match their supposed relationship.
+static bool prv_channels_synced(int16_t reading_main, int16_t reading_secondary,
+                                ThrottleStorage *storage) {
+  uint16_t numerator_main = prv_get_numerator(reading_main, THROTTLE_CHANNEL_MAIN, storage);
+
+  uint16_t expected_reading_secondary =
+      ((storage->calibration_data
+            ->line_of_best_fit[THROTTLE_CHANNEL_SECONDARY][THROTTLE_THRESH_MAX] -
+        storage->calibration_data
+            ->line_of_best_fit[THROTTLE_CHANNEL_SECONDARY][THROTTLE_THRESH_MIN]) *
+       numerator_main / (1 << 12)) +
+      storage->calibration_data->line_of_best_fit[THROTTLE_CHANNEL_SECONDARY][THROTTLE_THRESH_MIN];
+
+  return (abs(expected_reading_secondary - reading_secondary)) <=
```

Comment:
This will result in an underflow.

Reply:
They wouldn't, if they are both int16_t correct?